[quantization] State dict serialization of nnq.Linear

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24117 [quantization] JIT serialization for Conv2d
* #24116 [quantization] state_dict serialization for Conv2d + some bugfixes
* #24115 [quantization] Re-work Conv2d
* #24048 [quantization] JIT Serialization of nnq.Linear
* **#24047 [quantization] State dict serialization of nnq.Linear**
* #24046 [quantization] Simplified nnq.Linear class

Add `_{save_to,load_from}_state_dict` methods to `nnq.Linear` that explicitly deal with conversions from the Python attributes to the serialized state dict form

Differential Revision: [D16728346](https://our.internmc.facebook.com/intern/diff/D16728346)

Commit:
[quantization] State dict serialization of nnq.Linear

jerryzh168 commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -118,6 +119,45 @@ def forward(self, x):
+    # outside the process in which they were created, rather they should be derived
+    # from the QTensor weight.
+    def _save_to_state_dict(self, destination, prefix, keep_vars):
+        super()._save_to_state_dict(destination, prefix, keep_vars)
```
does `super()` work in python2?

jamesr66a replied:
nope, good catch

jerryzh168 commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -118,6 +119,45 @@ def forward(self, x):
+    def weight(self):
+        return torch.ops.quantized.fbgemm_linear_unpack(self._packed_weight)
+
+    def set_weight(self, w):
```
add a test for set_weight function as well? I think this is should be used in from_float

jerryzh168 replied:
line 188 we can probably use set_weight

Commit:
Update on "[quantization] State dict serialization of nnq.Linear"


Add `_{save_to,load_from}_state_dict` methods to `nnq.Linear` that explicitly deal with conversions from the Python attributes to the serialized state dict form

jerryzh168 commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -110,14 +111,56 @@ def __init__(self, in_features, out_features, bias_=True):
+        self.zero_point = int(state_dict[prefix + 'zero_point'])
+        state_dict.pop(prefix + 'zero_point')
+
+        super()._load_from_state_dict(state_dict, prefix, local_metadata, False,
```
..another one here.

jamesr66a replied:
I'm surprised flake8 doesn't catch these in the pre-commit hook...

jerryzh168 reviewed (approved):
LGTM, please fix the extra super()

Commit:
Update on "[quantization] State dict serialization of nnq.Linear"


Add `_{save_to,load_from}_state_dict` methods to `nnq.Linear` that explicitly deal with conversions from the Python attributes to the serialized state dict form

Commit:
Update on "[quantization] State dict serialization of nnq.Linear"


Add `_{save_to,load_from}_state_dict` methods to `nnq.Linear` that explicitly deal with conversions from the Python attributes to the serialized state dict form

Commit:
Update on "[quantization] State dict serialization of nnq.Linear"


Add `_{save_to,load_from}_state_dict` methods to `nnq.Linear` that explicitly deal with conversions from the Python attributes to the serialized state dict form

Commit:
Update on "[quantization] State dict serialization of nnq.Linear"


Add `_{save_to,load_from}_state_dict` methods to `nnq.Linear` that explicitly deal with conversions from the Python attributes to the serialized state dict form

Differential Revision: [D16728346](https://our.internmc.facebook.com/intern/diff/D16728346)

Commit:
Update on "[quantization] State dict serialization of nnq.Linear"


Add `_{save_to,load_from}_state_dict` methods to `nnq.Linear` that explicitly deal with conversions from the Python attributes to the serialized state dict form

Differential Revision: [D16728346](https://our.internmc.facebook.com/intern/diff/D16728346)
